pootle/apps/pootle_revision/utils.py

```diff
@@ -156,25 +156,38 @@ def get_revisions(self, parents, keys=None):
             key__in=keys or [""],
             object_id__in=parents)
 
-    def create_revisions(self, parents, keys=None):
-        new_revision = self.new_revision
-        for parent in parents:
-            for key in keys or [""]:
-                yield Revision(
-                    content_type_id=self.content_type_id,
-                    object_id=parent,
-                    key=key,
-                    value=new_revision)
-
     def update(self, keys=None):
         parents = list(self.parents.values_list("id", flat=True))
         revisions = self.get_revisions(parents, keys=keys)
-        # manually get the list of ids and delete those to prevent
-        # django race condition
-        revision_ids = list(revisions.values_list("id", flat=True))
-        revisions.filter(id__in=revision_ids).delete()
-        Revision.objects.bulk_create(
-            self.create_revisions(parents, keys=keys))
+        missing_revisions = []
+        existing_ids = []
+        revision_map = {
+            '%s-%s' % (x['object_id'], x['key']): x['id']
+            for x in revisions.values("id", "object_id", "key")}
+        for parent in parents:
+            for key in keys or [""]:
```

Comment:
I know this comes from previous code, but I don't see the need for having `keys or [""]` since if no `keys` is specified this will be iterated once resulting in an empty `key` which AFAICT will generate an empty `id`. I am not sure if it is ok to create a `Revision` without a `key`.